cmd/localstack/awsutil.go

```diff
@@ -196,6 +200,135 @@ func DownloadCodeArchive(url string) {
+	// Start listening for events.
+	go func(channel chan<- string) {
+		var watchedFolders []string
+		for {
+			select {
+			case event, ok := <-watcher.Events():
+				if !ok {
+					return
+				}
+				log.Debugln("FileWatcher got event: ", event)
+				if event.Has(fsnotify.Create) {
+					stat, err := os.Stat(event.Name)
+					if err != nil {
+						log.Errorln("Error stating event file: ", event.Name, err)
+					} else if stat.IsDir() {
+						subfolders := getSubFolders(event.Name)
+						for _, folder := range subfolders {
+							err = watcher.Add(folder)
+							watchedFolders = append(watchedFolders, folder)
+							if err != nil {
+								log.Errorln("Error watching folder: ", folder, err)
+							}
+						}
+					}
+					// remove in case of remove / rename (rename within the folder will trigger a separate create event)
+				} else if event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename) {
+					// remove all file watchers if it is in our folders list
+					toBeRemovedDirs, newWatchedFolders := getSubFoldersInList(event.Name, watchedFolders)
+					watchedFolders = newWatchedFolders
+					for _, dir := range toBeRemovedDirs {
+						err = watcher.Remove(dir)
+						if err != nil {
+							log.Warnln("Error removing path: ", event.Name, err)
+						}
+					}
+				}
+				channel <- event.Name
+			case err, ok := <-watcher.Errors():
+				if !ok {
+					log.Println("error:", err)
+					return
+				}
+				log.Println("error:", err)
+			}
+		}
+	}(changeChannel)
+
+	// debouncer to limit restarts
+	go func(channel <-chan string, duration time.Duration) {
+		timer := time.NewTimer(duration)
+		// immediately stop the timer, since we do not want to reload right at the startup
+		if !timer.Stop() {
+			// we have to drain the channel in case the timer already fired
+			<-timer.C
+		}
+		for {
+			select {
+			case _, more := <-channel:
+				if !more {
+					timer.Stop()
+					return
+				}
+				timer.Reset(duration)
+			case <-timer.C:
+				log.Println("Resetting environment...")
+				server.Reset("HotReload", 2000)
+			}
+		}
+
+	}(changeChannel, defaultDuration)
+
+	// Add all target paths and subfolders
+	for _, targetPath := range targetPaths {
+		subfolders := getSubFolders(targetPath)
+		log.Infoln("Subfolders: ", subfolders)
+		for _, target := range subfolders {
+			err = watcher.Add(target)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+	<-done
+	log.Infoln("Closing down filewatcher.")
+
+}
+
+func getSubFolders(dirPath string) []string {
+	var subfolders []string
+	err := filepath.WalkDir(dirPath, func(path string, d fs.DirEntry, err error) error {
+		if err == nil && d.IsDir() {
+			subfolders = append(subfolders, path)
+		}
+		return err
+	})
+	if err != nil {
+		log.Errorln("Error listing directory contents: ", err)
+		return subfolders
+	}
+	return subfolders
+}
+
+func getSubFoldersInList(prefix string, pathList []string) (old_folders []string, new_folders []string) {
+	for _, item := range pathList {
```

Comment:
nit: `item` is a bit too generic here. 

```suggestion
	for _, pathItem := range pathList {
```